### PR TITLE
Feat(enhancement): consider afterburners separately and allow them to have cooldown and duration

### PR DESCRIPTION
**Pull request by Hurleveur**
**Feature:** This PR implements the feature request by @Saugia , detailed and discussed in discord

## Feature Details
One can add `afterburner cooldown` and `afterburner endurance` to any afterburner.
If that is done, the afterburner will only fire so long as the duration specifies, and then enter a cooldown mode.
If the duration is of 10 (seconds) and the cooldown of 100, that means one can use the afterburner 10 seconds before having to wait the full 100, or use it 5 seconds, wait 5, use it 5 etc.
Meaning one is incentivised to not push the afterburner to its limit.

In order to make this work, afterburners are now considered separately instead of just all being mixed and applied on the ship at once.
That means that in the future it'll be easier to separate them to for instance only use one of them.

## UI Screenshots
N/A

## Usage Examples
N/A

## Testing Done
Did not test yet

### Automated Tests Added
N/A

## Performance Impact
N/A

### CODE REVIEW

**Comment by Hurleveur** [comment]
Later on I can see this being expanded to be used for more than just afterburners actually.
When we do we should probably rename it too, but rn having it be generic and used only for afterburners would be weird.

**Comment by Hurleveur** [comment]
Is macos failing because my computer changed the xcode ending line style when running it?
Idk why it wouldnt be compatible with linux ending lines honestly I'm confused
~~That'd make it even worse than what I thought it was~~

**Comment by Saugia** [comment]
I had requested this for a unique afterburner for the Gegno (and for implementation in general for those working on plug-ins, etc.). I was looking to make a *Burstfire Afterburner* - one that has a burst of propulsion for a second or so, then has a cooldown of several seconds. Different than that of a constant afterburner, more situational, etc.. If we could get this to work somehow, that would be great!

**Comment by Hurleveur** [comment]
Well it does work: I tested it manually and the xcodeproj issue was fixed thx to quyykk noticing a weird thing.

**Comment by Hurleveur** [comment]
I swear I hate macos.
Each time a conflict and I cant even just accept both versions and have it work.
I'll do this manually again, btw could we make it happen automatically?

File: data/tooltips.txt [hunk 1]
```diff
@@ -41,6 +41,12 @@ tip "afterburner heat:"
 tip "afterburner thrust:"
 	`Thrust produced by this afterburner.`
 
+tip "afterburner cooldown:"
+	`The takes it times for the afterburner to be available again.`
```
Comment (Koranir) [inline data/tooltips.txt:45]:
```suggestion
	`The time it takes in seconds for the afterburner to be available again after the duration ends.`
```

File: data/tooltips.txt [hunk 2]
```diff
@@ -41,6 +41,12 @@ tip "afterburner heat:"
...
+
+tip "afterburner duration:"
+	`The maximum length of time this afterburner can be used for, in a continuous manner, before requiring reload.`
```
Comment (Koranir) [inline data/tooltips.txt:48]:
```suggestion
	`The maximum duration this afterburner can be used for. If this time is exceeded, this afterburner goes into a cooldown period.`
```

**Review by Koranir — COMMENTED** [review]
Making things a bit clearer

File: source/OutfitInfoDisplay.cpp [hunk 3]
```diff
@@ -121,6 +121,9 @@ namespace {
 		{"afterburner burn", 2},
 		{"afterburner slowing", 2},
 		{"afterburner disruption", 2},
+		{"afterburner duration", 0},
```
Comment (Koranir) [inline source/OutfitInfoDisplay.cpp:124]:
```suggestion
		{"afterburner duration", 6},
```

File: source/OutfitInfoDisplay.cpp [hunk 4]
```diff
@@ -121,6 +121,9 @@ namespace {
...
+		{"afterburner cooldown", 0},
```
Comment (Koranir) [inline source/OutfitInfoDisplay.cpp:125]:
```suggestion
		{"afterburner cooldown", 6},
```

**Review by Koranir — COMMENTED** [review]
Divides by 60 instead of multiplying.

**Comment by Hurleveur** [comment]
Oh right I'll be honest idk much about those scaling factors, so ty for helping me with those!

File: source/OutfitInfoDisplay.cpp [hunk 5]
```diff
@@ -121,6 +121,9 @@ namespace {
 		{"afterburner burn", 2},
 		{"afterburner slowing", 2},
 		{"afterburner disruption", 2},
+		{"afterburner duration", 6},
```
Comment (Koranir) [inline source/OutfitInfoDisplay.cpp:124]:
```suggestion
		{"afterburner duration", 5},
```

File: source/OutfitInfoDisplay.cpp [hunk 6]
```diff
@@ -121,6 +121,9 @@ namespace {
...
+		{"afterburner cooldown", 6},
```
Comment (Koranir) [inline source/OutfitInfoDisplay.cpp:125]:
```suggestion
		{"afterburner cooldown", 5},
```

**Review by Koranir — COMMENTED** [review]
Whoops, I made a huge mistake, vectors start at 0 not 1. Sorry :P

File: source/AfterburnerUsage.h [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+#include "Outfit.h"
+
+
+
+// Class used to manage afterburners and their cooldowns/usages on ships.
+class AfterburnerUsage {
+public:
+	// Constructor
+	AfterburnerUsage(const Outfit &outfit);
```
Comment (RisingLeaf) [inline source/AfterburnerUsage.h:27]: Always good to have a default constructor, just in case.

File: source/Outfit.cpp [hunk 8]
```diff
@@ -435,6 +435,13 @@ const Dictionary &Outfit::Attributes() const
 
 
 
+bool Outfit::IsAfterburner() const
```
Comment (RisingLeaf) [inline source/Outfit.cpp:438]: Because outfits can be combined maybe more like hasafterburner()?

File: source/AfterburnerUsage.h (hunk 7 above)
Comment (Hurleveur) [reply source/AfterburnerUsage.h:27]: Idk, look at AI it does not have a default constructor. This makes no sense without an attached outfit, and btw the default constructor is there if you dont delete it I think?

**Comment by Zitchas** [comment]
Is it possible to have an afterburner with a rocket-style functionality? As in, once you hit afterburner, it is *on* until it reaches the end of its duration? so "afterburner duration" 5 means that when you tap the afterburner key, you spend the next 5 seconds with the afterburner pushing at max, even if you let go of the afterburner key. This is what I was originally thinking of in terms of "afterburner duration" in the conversation the other day.

**Comment by Hurleveur** [comment]
I must have missed that conversation.
That sounds like a proper Ork thruster for my Unfettered boys too.... dont worry I wont keep rights on it x)
I won't add that to this PR but it could *easily* be done as a follow up, by just adding a new attribute and doing minor changes.

Now, for the name, I just wrote it quickly, so I'll change it. 
Now, idk about `afterburner endurance`... I feel like we can get a better name.

File: data/tooltips.txt [hunk 9]
```diff
@@ -41,6 +41,12 @@ tip "afterburner heat:"
 tip "afterburner thrust:"
 	`Thrust produced by this afterburner.`
 
+tip "afterburner cooldown:"
+	`The time it takes in seconds for the afterburner to be available again after the duration ends.`
```
Comment (quyykk) [inline data/tooltips.txt:45]: It seems like the time isn't in seconds but in frames from what I can tell? Either way, it should be in frames like everything else.

File: source/AfterburnerUsage.cpp [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+this program. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "AfterburnerUsage.h"
+
+
+
+AfterburnerUsage::AfterburnerUsage(const Outfit &outfit) :
+	afterburner(outfit), baseDuration(outfit.Attributes().Get("afterburner duration")),
+	baseCooldown(outfit.Attributes().Get("afterburner cooldown"))
```
Comment (quyykk) [inline source/AfterburnerUsage.cpp:22]: Pretty sure the durations aren't updated if you uninstall/install afterburners?

Comment (Hurleveur) [reply source/AfterburnerUsage.cpp:22]: Oh that isnt like buying then.... I'll look into it

Comment (Hurleveur) [reply source/AfterburnerUsage.cpp:22]: AddOutfit isnt used for that? that would be weird

Comment (quyykk) [reply source/AfterburnerUsage.cpp:22]: Are you sure you did? The only place you add to the afterburner list is in Ship::FinishLoading.

Comment (Hurleveur) [reply source/AfterburnerUsage.cpp:22]: in AddOutfit too?

**Comment by Hurleveur** [comment]
I'm going to be going on a trip for 3 months, meaning I will be much less active on my PRs.
That said, they all are as ready as I could make them, which is why I'm not closing them or marking them as waiting for op.

I won't be able to test and make changes fast, but feel free to post comments and especially to do the needed changes, I truly don't mind. In fact I'd like that.

This PR is pretty simple code wise, and lays the groundworks for deactivating some outfits - right now it's only outfitters as we need them, but we can eventually do more with it.
The integration test for afterburner only running should make sure this doesnt break anything, and I tested it manually extensively.

**Comment by Amazinite** [comment]
I'm not really a fan of implementing this in such a way where this individual outfit acting behavior only works for afterburners. Afterburners are not the only outfit where it would make sense to have them act individually, and there are outfits which people have discussed potentially creating that would almost only make sense to have act in an individual manner. As such, I think that any attempt to make one type of outfit act individually should be implemented in a way that can be easily expanded to other types of outfits.

**Pull request closed without merging**